libcontainer/devices/device.go

```diff
@@ -24,6 +24,9 @@ type Device struct {
 
 	// Gid of the device.
 	Gid uint32 `json:"gid"`
+
+	// Is Default Device
+	IsDefault bool `json:"is_default"`
```

Comment:
This struct is stored on disk and so, on upgrade, any existing containers will have a JSON config that has ever device be "non-default". It seems like this is only used during specconv, so maybe this should be a hidden JSON field rather than being stored on-disk?

(As an aside, we really should have tests to make sure we never break this again -- Docker 17.06.1 was a complete nightmare. But in this particular case, nothing would break but the behaviour would still not be ideal.)

Reply:
> It seems that this flag is only used for an internal check (and the on-disk flag is never checked)

Afaik, the on-disk flag is used when updating. `CreateDefaultDevicesCgroups` is now called from `update.go` (see `updateCommand`).

From what I remember this was necessary to "clear" allowed devices. Otherwise, the user would have to pass the default devices to get back to default state.

> This struct is stored on disk and so, on upgrade, any existing containers will have a JSON config that has ever device be "non-default".

Hm, this is a problem, can we migrate the data?

Reply:
I've reviewed this again, the reason it is required is because default allowed devices get special treatment for via `CreateCgroupConfig`. To do this, we need to understand which devices have been added through spec, and which have been added from  `AllowedDevices`. This used to be done in a single function `createDevices`. With commit https://github.com/opencontainers/runc/pull/3402/commits/390ac2b10b5d1b46479d1bb1ca64a99d72cbd972 de-duplicates this logic so it can be used for updates.

There is no need to store this flag in JSON, so I think we can use ``json:"-"`` which we use in other places too.

Reply:
It turns out, the `is_default` needs to be stored in the container state to successfully "re-duplicate" on update.